tensorrt_llm/_torch/pyexecutor/resource_manager.py

```diff
@@ -501,6 +565,150 @@ def get_kv_cache_stats(self):
+                    "max_attention_window_vec cannot be empty if there are local layers."
+                )
+            return {
+            }  # Return an empty dict if no local layers or if somehow vec is empty and no layers.
+
+        # Treat max_attention_window_vec as a repeating pattern.
+        # This also correctly handles the case where len(self.max_attention_window_vec) == 1.
+        pattern_len = len(self.max_attention_window_vec)
+        for local_layer_idx in range(self.num_local_layers):
+            window_size = self.max_attention_window_vec[local_layer_idx %
+                                                        pattern_len]
+            window_size_to_layers_map[window_size].append(local_layer_idx)
+        return window_size_to_layers_map
+
+    def calculate_max_num_blocks_from_cpp(
+            self,
+            kv_cache_config: KvCacheConfigCpp,
+            model_config: ModelConfig,
+            extra_cost_memory: int = 0) -> dict[int, tuple[int, int]]:
+        """
+        This function is a wrapper of KVCacheManagerCpp.calculate_max_num_blocks.
+        The final goal is to switch to the C++ implementation of calculate_max_num_blocks.
+        Currently, this function is added to support *ONLY* VSWA.
+
+        Args:
+            kv_cache_config: The KV cache configuration object.
+            model_config: The model configuration object.
+            extra_cost_memory: Extra memory in bytes to exclude from available memory.
+
+        Returns:
+            A dict of (max_attention_window, (blocks_in_primary_pool, blocks_in_secondary_pool)).
+        """
+
+        # Construct WorldConfig from self.mapping
+        world_config_cpp = WorldConfig(
+            tensor_parallelism=self.mapping.tp_size,
+            pipeline_parallelism=self.mapping.pp_size,
+            rank=self.mapping.rank,
+            gpus_per_node=self.mapping.gpus_per_node)
+
+        window_size_to_layers = self._get_window_size_to_layers()
+        logger.debug(f"window_size_to_layers: {window_size_to_layers}")
+
+        free_mem, total_mem = torch.cuda.mem_get_info()
+        primary_pool_memory_bytes = free_mem
+        secondary_pool_memory_bytes = 0
+        logger.debug(
+            f"primary_pool_memory_bytes is set to {primary_pool_memory_bytes/1024**3}GB, \nsecondary_pool_memory_bytes is set to {secondary_pool_memory_bytes/1024**3}GB"
+        )
+
+        blocks_per_window = KVCacheManagerCpp.calculate_max_num_blocks(
+            config=kv_cache_config,
+            is_cross_attention=False,  #TODO: support cross attention
+            dtype=self.dtype,
+            model_config=model_config,
+            world_config=world_config_cpp,
+            window_size_to_layers=window_size_to_layers,
+            allotted_primary_mem_bytes=primary_pool_memory_bytes,
+            allotted_secondary_mem_bytes=secondary_pool_memory_bytes,
+            extra_cost_memory=extra_cost_memory,
+            kvFactor=self.kv_factor,
```

Comment:
Not your fault, but I would just fix `kvFactor` to be kv_factor in the bindings at this point